### PR TITLE
Deleting deployer pods for failed deployment before retrying

### PR DESCRIPTION
**Pull request by abhgupta**
Closes https://github.com/openshift/origin/issues/2558

### CODE REVIEW

File: pkg/cmd/cli/cmd/deploy.go [hunk 1]
```diff
@@ -250,6 +265,18 @@ func (c *retryDeploymentCommand) retry(config *deployapi.DeploymentConfig, out i
 		return fmt.Errorf(message)
 	}
 
+	// Delete the deployer pod as well as the deployment hooks pods, if any
+	pods, err := c.client.ListDeployerPodsFor(config.Namespace, deploymentName)
+	if err != nil {
+		return fmt.Errorf("Failed to list deployer/hook pods for the failed deployment: %v", err)
```
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Report the deployment namespace/name in the message.

File: pkg/cmd/cli/cmd/deploy.go [hunk 2]
```diff
@@ -250,6 +265,18 @@ func (c *retryDeploymentCommand) retry(config *deployapi.DeploymentConfig, out i
...
+	}
+	for _, pod := range pods.Items {
+		err := c.client.DeletePod(&pod)
+		if err != nil {
+			return fmt.Errorf("Failed to delete deployer/hook pod %s for the failed deployment: %v", pod.Name, err)
```
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Report the deployment namespace/name in the message.

File: pkg/cmd/cli/cmd/deploy_test.go [hunk 3]
```diff
@@ -131,8 +131,14 @@ func TestCmdDeploy_latestLookupError(t *testing.T) {
...
 // TestCmdDeploy_retryOk ensures that a failed deployment can be retried.
 func TestCmdDeploy_retryOk(t *testing.T) {
+	deletedPods := []string{}
+	existingDeployerPods := []kapi.Pod{}
+
 	config := deploytest.OkDeploymentConfig(1)
 	existingDeployment := deploymentFor(config, deployapi.DeploymentStatusFailed)
+	existingDeployerPods = append(existingDeployerPods, kapi.Pod{ObjectMeta: kapi.ObjectMeta{Name: "prehook"}})
+	existingDeployerPods = append(existingDeployerPods, kapi.Pod{ObjectMeta: kapi.ObjectMeta{Name: "posthook"}})
+	existingDeployerPods = append(existingDeployerPods, kapi.Pod{ObjectMeta: kapi.ObjectMeta{Name: "deployerpod"}})
```
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy_test.go]: Nit: you could just declare all this inline with the array declaration.

File: pkg/cmd/cli/cmd/deploy_test.go [hunk 4]
```diff
@@ -160,6 +173,11 @@ func TestCmdDeploy_retryOk(t *testing.T) {
 		t.Fatalf("expected updated config")
 	}
 
+	sort.Strings(deletedPods)
+	if !reflect.DeepEqual(deletedPods, []string{"deployerpod", "posthook", "prehook"}) {
```
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy_test.go]: Using `util.StringSet` instead of an array might make this easier.

File: pkg/cmd/cli/cmd/deploy.go [hunk 5]
```diff
@@ -36,7 +37,7 @@ type DeployOptions struct {
 }
 
 const (
-	deployLong = `View, start and restart deployments.
+	deployLong = `View, start, cancel, and retry deployments.
```
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Should we point out in the `retry` help that the command will be destructive to existing pods and that any pod/container info the user needs should be captured prior to retrying?

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (abhgupta) [inline pkg/cmd/cli/cmd/deploy.go]: The rest of the messages in this file do not have use this format and just use the depoyment number. I have made these messages consistent with the rest.

File: pkg/cmd/cli/cmd/deploy.go (hunk 2 above)
Comment (abhgupta) [inline pkg/cmd/cli/cmd/deploy.go]: The rest of the messages in this file do not have use this format and just use the depoyment number. I have made these messages consistent with the rest.

File: pkg/cmd/cli/cmd/deploy_test.go (hunk 3 above)
Comment (abhgupta) [inline pkg/cmd/cli/cmd/deploy_test.go]: Done.

File: pkg/cmd/cli/cmd/deploy_test.go (hunk 4 above)
Comment (abhgupta) [inline pkg/cmd/cli/cmd/deploy_test.go]: util.StringSet would require mutiple (bi-directional) calls to Difference to achieve this. Using sort and string slices is just simpler in this case.

File: pkg/cmd/cli/cmd/deploy.go (hunk 5 above)
Comment (abhgupta) [inline pkg/cmd/cli/cmd/deploy.go]: Yes, have to think of something that is not too scary.

**Comment by abhgupta** [comment]
@ironcladlou feedback implemented

**Comment by ironcladlou** [comment]
LGTM [merge]

**Comment by ironcladlou** [comment]
cc @thoraxe @sdodson: this is a critical fix made in the past week that you guys might need to be aware of.

**Pull request merged**